// @thumbtack/thumbprint-react/dist/es/components/Link/index.js
import { extends as _extends } from '../../_virtual/_rollupPluginBabelHelpers.js';
import React from 'react';
import warning from 'warning';
import Themed from '../UIAction/themed.js';
import Plain from '../UIAction/plain.js';

var getCommonLinkProps = function getCommonLinkProps(props) {
  warning( // If `onClick` prop exists then `to` must also exist.
  !props.onClick || props.to && props.onClick, 'The `onClick` prop in the Thumbprint `Link` and `ThemedLink` components should only be used when the `to` prop is also provided. Consider using the `Button` component instead.');
  return {
    to: props.to,
    onClick: props.onClick,
    target: props.target,
    shouldOpenInNewTab: props.shouldOpenInNewTab,
    isDisabled: props.isDisabled,
    children: props.children,
    accessibilityLabel: props.accessibilityLabel,
    dataTestId: props.dataTestId,
    dataTest: props.dataTest
  };
};
/**
 * Anchor link that renders as text.
 */


var Link = React.forwardRef(function (_ref, ref) {
  var to = _ref.to,
      onClick = _ref.onClick,
      target = _ref.target,
      _ref$shouldOpenInNewT = _ref.shouldOpenInNewTab,
      shouldOpenInNewTab = _ref$shouldOpenInNewT === void 0 ? false : _ref$shouldOpenInNewT,
      _ref$isDisabled = _ref.isDisabled,
      isDisabled = _ref$isDisabled === void 0 ? false : _ref$isDisabled,
      children = _ref.children,
      accessibilityLabel = _ref.accessibilityLabel,
      dataTest = _ref.dataTest,
      dataTestId = _ref.dataTestId,
      _ref$theme = _ref.theme,
      theme = _ref$theme === void 0 ? 'primary' : _ref$theme,
      iconLeft = _ref.iconLeft,
      iconRight = _ref.iconRight;
  return React.createElement(Plain, _extends({}, getCommonLinkProps({
    to: to,
    onClick: onClick,
    target: target,
    shouldOpenInNewTab: shouldOpenInNewTab,
    isDisabled: isDisabled,
    children: children,
    accessibilityLabel: accessibilityLabel,
    dataTest: dataTest,
    dataTestId: dataTestId
  }), {
    theme: theme,
    iconLeft: iconLeft,
    iconRight: iconRight,
    ref: ref
  }));
});

/**
 * Anchor link that visually looks like a button.
 */
var ThemedLink = React.forwardRef(function (_ref2, ref) {
  var to = _ref2.to,
      onClick = _ref2.onClick,
      _ref2$shouldOpenInNew = _ref2.shouldOpenInNewTab,
      shouldOpenInNewTab = _ref2$shouldOpenInNew === void 0 ? false : _ref2$shouldOpenInNew,
      target = _ref2.target,
      _ref2$isDisabled = _ref2.isDisabled,
      isDisabled = _ref2$isDisabled === void 0 ? false : _ref2$isDisabled,
      children = _ref2.children,
      accessibilityLabel = _ref2.accessibilityLabel,
      dataTest = _ref2.dataTest,
      dataTestId = _ref2.dataTestId,
      icon = _ref2.icon,
      iconRight = _ref2.iconRight,
      _ref2$theme = _ref2.theme,
      theme = _ref2$theme === void 0 ? 'primary' : _ref2$theme,
      _ref2$size = _ref2.size,
      size = _ref2$size === void 0 ? 'large' : _ref2$size,
      _ref2$width = _ref2.width,
      width = _ref2$width === void 0 ? 'auto' : _ref2$width;
  return React.createElement(Themed, _extends({}, getCommonLinkProps({
    to: to,
    onClick: onClick,
    shouldOpenInNewTab: shouldOpenInNewTab,
    target: target,
    isDisabled: isDisabled,
    children: children,
    accessibilityLabel: accessibilityLabel,
    dataTest: dataTest,
    dataTestId: dataTestId
  }), {
    icon: icon,
    iconRight: iconRight,
    theme: theme,
    size: size,
    width: width,
    ref: ref
  }));
});

export default Link;
export { ThemedLink };
