// @thumbtack/thumbprint-react/dist/es/components/UIAction/get-anchor-props.js
import isThumbtackUrl from './is-thumbtack-url.js';

var getRel = function getRel(url, target) {
  var shouldOpenInNewTab = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;

  if (shouldOpenInNewTab || target === '_blank') {
    if (isThumbtackUrl(url)) {
      // There are performance benefits of adding `rel="noopener"`.
      // https://jakearchibald.com/2016/performance-benefits-of-rel-noopener/
      return 'noopener';
    } // https://mathiasbynens.github.io/rel-noopener/
    // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-target-blank.md


    return 'noopener noreferrer';
  }

  return undefined;
};

/**
 * Enable plain and themed `<a>` elements to share the same props.
 */
var getAnchorProps = function getAnchorProps(_ref) {
  var isDisabled = _ref.isDisabled,
      shouldOpenInNewTab = _ref.shouldOpenInNewTab,
      to = _ref.to,
      onClick = _ref.onClick,
      target = _ref.target;
  return {
    onClick: onClick,
    href: isDisabled ? undefined : to,
    target: target || (shouldOpenInNewTab ? '_blank' : '_self'),
    // NOTE use `noopener` even for internal links in new tabs to prevent potential slowdown of
    // new tab https://mathiasbynens.github.io/rel-noopener/
    rel: getRel(to, target, shouldOpenInNewTab)
  };
};

export default getAnchorProps;
